File: web/src/lib/ports/transient-queue.ts
export interface EncryptedPayload {
  ciphertext: string;
  ratchetHeader?: string;
}

export interface QueuedMessage {
  messageId: string;
  payload: EncryptedPayload;
  enqueuedAt: number;
}

export interface TransientQueue {
  enqueue(recipientDeviceId: string, payload: EncryptedPayload): Promise<void>;
  dequeue(deviceId: string, limit?: number): Promise<QueuedMessage[]>;
  acknowledge(deviceId: string, messageIds: string[]): Promise<void>;
  getQueueDepth(deviceId: string): Promise<number>;
}

File: engine/lib/src/ports/transient_queue.dart
class QueuedMessage {
  final String messageId;
  final Map<String, dynamic> payload;
  final int enqueuedAt;
  const QueuedMessage({
    required this.messageId,
    required this.payload,
    required this.enqueuedAt,
  });
}

abstract class TransientQueue {
  Future<void> enqueue(String recipientDeviceId, Map<String, dynamic> payload);
  Future<List<QueuedMessage>> dequeue(String deviceId, {int limit = 100});
  Future<void> acknowledge(String deviceId, List<String> messageIds);
  Future<int> getQueueDepth(String deviceId);
}
